module/zfs/arc.c

```diff
@@ -9110,9 +9111,9 @@ l2arc_apply_transforms(spa_t *spa, arc_buf_hdr_t *hdr, uint64_t asize,
 		}
 		ASSERT3U(psize, <=, HDR_GET_PSIZE(hdr));
 		if (psize < asize)
-			memset((char *)tmp + psize, 0, asize - psize);
+			memset((char *)tmp + psize, 0, bufsize - psize);
 		psize = HDR_GET_PSIZE(hdr);
-		abd_return_buf_copy(cabd, tmp, size);
+		abd_return_buf_copy(cabd, tmp, bufsize);
 		to_write = cabd;
```

Comment:
Nit: rather then using a "goto encrypt" here, I think it would be cleaner to just turn this section into an `else` clause something like:
```
} else {
    /* pad out to asize bytes with zeros */
    ASSERT3U(csize, <=, psize);
    if (csize < asize)
        memset((char *)tmp + csize, 0, asize - csize);
    abd_return_buf_copy(cabd, tmp, bufsize);
    to_write = cabd;
}
```
Note, this would allow you to also move the common `to_write = cabd` from each clause down below the if-statement.